// @rn-primitives/tooltip/dist/tooltip.mjs
'use client'
import {
  __objRest,
  __spreadValues
} from "./chunk-X44HCPWV.mjs";

// src/tooltip.tsx
import { useAugmentedRef, useRelativePosition } from "@rn-primitives/hooks";
import { Portal as RNPPortal } from "@rn-primitives/portal";
import * as Slot from "@rn-primitives/slot";
import * as React from "react";
import {
  BackHandler,
  Pressable as Pressable2,
  View as View2
} from "react-native";
var RootContext = React.createContext(null);
var Root = React.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      asChild,
      delayDuration: _delayDuration,
      skipDelayDuration: _skipDelayDuration,
      disableHoverableContent: _disableHoverableContent,
      onOpenChange: onOpenChangeProp
    } = _b, viewProps = __objRest(_b, [
      "asChild",
      "delayDuration",
      "skipDelayDuration",
      "disableHoverableContent",
      "onOpenChange"
    ]);
    const nativeID = React.useId();
    const [triggerPosition, setTriggerPosition] = React.useState(null);
    const [contentLayout, setContentLayout] = React.useState(null);
    const [open, setOpen] = React.useState(false);
    function onOpenChange(value) {
      setOpen(value);
      onOpenChangeProp == null ? void 0 : onOpenChangeProp(value);
    }
    const Component = asChild ? Slot.View : View2;
    return /* @__PURE__ */ React.createElement(
      RootContext.Provider,
      {
        value: {
          open,
          onOpenChange,
          contentLayout,
          nativeID,
          setContentLayout,
          setTriggerPosition,
          triggerPosition
        }
      },
      /* @__PURE__ */ React.createElement(Component, __spreadValues({ ref }, viewProps))
    );
  }
);
Root.displayName = "RootNativeTooltip";
function useTooltipContext() {
  const context = React.useContext(RootContext);
  if (!context) {
    throw new Error("Tooltip compound components cannot be rendered outside the Tooltip component");
  }
  return context;
}
var Trigger = React.forwardRef(
  (_a, ref) => {
    var _b = _a, { asChild, onPress: onPressProp, disabled = false } = _b, props = __objRest(_b, ["asChild", "onPress", "disabled"]);
    const { open, onOpenChange, setTriggerPosition } = useTooltipContext();
    const augmentedRef = useAugmentedRef({
      ref,
      methods: {
        open: () => {
          var _a2;
          onOpenChange(true);
          (_a2 = augmentedRef.current) == null ? void 0 : _a2.measure((_x, _y, width, height, pageX, pageY) => {
            setTriggerPosition({ width, pageX, pageY, height });
          });
        },
        close: () => {
          setTriggerPosition(null);
          onOpenChange(false);
        }
      }
    });
    function onPress(ev) {
      var _a2;
      if (disabled) return;
      (_a2 = augmentedRef.current) == null ? void 0 : _a2.measure((_x, _y, width, height, pageX, pageY) => {
        setTriggerPosition({ width, pageX, pageY, height });
      });
      const newValue = !open;
      onOpenChange(newValue);
      onPressProp == null ? void 0 : onPressProp(ev);
    }
    const Component = asChild ? Slot.Pressable : Pressable2;
    return /* @__PURE__ */ React.createElement(
      Component,
      __spreadValues({
        ref: augmentedRef,
        "aria-disabled": disabled != null ? disabled : void 0,
        role: "button",
        onPress,
        disabled: disabled != null ? disabled : void 0
      }, props)
    );
  }
);
Trigger.displayName = "TriggerNativeTooltip";
function Portal({ forceMount, hostName, children }) {
  const value = useTooltipContext();
  if (!value.triggerPosition) {
    return null;
  }
  if (!forceMount) {
    if (!value.open) {
      return null;
    }
  }
  return /* @__PURE__ */ React.createElement(RNPPortal, { hostName, name: `${value.nativeID}_portal` }, /* @__PURE__ */ React.createElement(RootContext.Provider, { value }, children));
}
var Overlay = React.forwardRef(
  (_a, ref) => {
    var _b = _a, { asChild, forceMount, onPress: OnPressProp, closeOnPress = true } = _b, props = __objRest(_b, ["asChild", "forceMount", "onPress", "closeOnPress"]);
    const { open, onOpenChange, setContentLayout, setTriggerPosition } = useTooltipContext();
    function onPress(ev) {
      if (closeOnPress) {
        setTriggerPosition(null);
        setContentLayout(null);
        onOpenChange(false);
      }
      OnPressProp == null ? void 0 : OnPressProp(ev);
    }
    if (!forceMount) {
      if (!open) {
        return null;
      }
    }
    const Component = asChild ? Slot.Pressable : Pressable2;
    return /* @__PURE__ */ React.createElement(Component, __spreadValues({ ref, onPress }, props));
  }
);
Overlay.displayName = "OverlayNativeTooltip";
var Content = React.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      asChild = false,
      forceMount,
      align = "center",
      side = "top",
      sideOffset = 0,
      alignOffset = 0,
      avoidCollisions = true,
      onLayout: onLayoutProp,
      insets,
      style,
      disablePositioningStyle
    } = _b, props = __objRest(_b, [
      "asChild",
      "forceMount",
      "align",
      "side",
      "sideOffset",
      "alignOffset",
      "avoidCollisions",
      "onLayout",
      "insets",
      "style",
      "disablePositioningStyle"
    ]);
    const {
      open,
      onOpenChange,
      nativeID,
      contentLayout,
      setContentLayout,
      setTriggerPosition,
      triggerPosition
    } = useTooltipContext();
    React.useEffect(() => {
      const backHandler = BackHandler.addEventListener("hardwareBackPress", () => {
        setTriggerPosition(null);
        setContentLayout(null);
        onOpenChange(false);
        return true;
      });
      return () => {
        setContentLayout(null);
        backHandler.remove();
      };
    }, []);
    const positionStyle = useRelativePosition({
      align,
      avoidCollisions,
      triggerPosition,
      contentLayout,
      alignOffset,
      insets,
      sideOffset,
      side,
      disablePositioningStyle
    });
    function onLayout(event) {
      setContentLayout(event.nativeEvent.layout);
      onLayoutProp == null ? void 0 : onLayoutProp(event);
    }
    if (!forceMount) {
      if (!open) {
        return null;
      }
    }
    const Component = asChild ? Slot.View : View2;
    return /* @__PURE__ */ React.createElement(
      Component,
      __spreadValues({
        ref,
        role: "tooltip",
        nativeID,
        "aria-modal": true,
        style: [positionStyle, style],
        onLayout,
        onStartShouldSetResponder
      }, props)
    );
  }
);
Content.displayName = "ContentNativeTooltip";
function onStartShouldSetResponder() {
  return true;
}
export {
  Content,
  Overlay,
  Portal,
  Root,
  Trigger
};
